Multicast/udpcast-20040531/udp-receiver.1
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.32
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "UDP-RECEIVER 1"
.TH UDP-RECEIVER 1 "July 27, 2007" "current" "Udpcast"
.SH "NAME"
udp\-receiver \- receive files broadcast by udp\-sender
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&./udp\-receiver [\-\-file \fIfile\fR] [\-\-pipe \fIpipe\fR] [\-\-portbase \fIportbase\fR] [\-\-interface \fInet-interface\fR] [\-\-log \fIfile\fR] [\-\-ttl \fItime-to-live\fR] [\-\-mcast\-all\-addr \fImcast-all-address\fR] [\-\-nokbd] [\-\-exitWait milliseconds]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\f(CW\*(C`Udp\-receiver\*(C'\fR is used to receive files sent by
\&\f(CW\*(C`udp\-sender\*(C'\fR (for instance a disk image).
.SH "OPTIONS"
.IX Header "OPTIONS"
.Sh "Basic options"
.IX Subsection "Basic options"
.IP "\-\-file \fIfile\fR" 4
.IX Item "--file file"
Writes received data to \fIfile\fR. If this parameter is not
supplied, received data is written to stdout instead.
.IP "\-\-pipe \fIcommand\fR" 4
.IX Item "--pipe command"
Sends data through \fIpipe\fR after receiving it. This is useful
for decompressing the data, or for filling in unused filesystem blocks
that may have been stripped out by udp\-sender. The \fIcommand\fR gets
a direct handle on the output file or device, and thus may seek inside
it, if needed. \f(CW\*(C`Udpcast\*(C'\fR itself also keeps a handle on the
file, which is used for an informational progress display. The
\&\fIcommand\fR's stdin is a pipe from udp\-receiver. Example:
\&\f(CW\*(C`udp\-receiver \-p "gzip \-dc"\*(C'\fR
.IP "\-\-log \fIfile\fR" 4
.IX Item "--log file"
Logs some stuff into \fIfile\fR.
.IP "\-\-nosync" 4
.IX Item "--nosync"
Do not open target in synchronous mode. Useful when dumping to an
image file on a mounted filesystem, rather than to a raw partition.
.IP "\-\-nokbd" 4
.IX Item "--nokbd"
Do not read start signal from keyboard, and do not display any
message telling the user to press any key to start.
.Sh "Networking options"
.IX Subsection "Networking options"
.IP "\-\-portbase \fIportbase\fR" 4
.IX Item "--portbase portbase"
Default ports to use for udpcast. Two ports are used:
\&\fIportbase\fR and \fIportbase+1\fR . Thus, \fIPortbase\fR must be
even. Default is \f(CW9000\fR. The same \fIportbase\fR must be
specified for both \f(CW\*(C`udp\-sender\*(C'\fR and \f(CW\*(C`udp\-receiver\*(C'\fR.
.IP "\-\-interface \fIinterface\fR" 4
.IX Item "--interface interface"
Network interface used to send out the data. Default is \f(CW\*(C`eth0\*(C'\fR
.IP "\-\-ttl \fIttl\fR" 4
.IX Item "--ttl ttl"
Time to live for connection request packet (by default connection
request is broadcast to the \s-1LAN\s0's broadcast address. If ttl is set,
the connection request is multicast instead to \f(CW224.0.0.1\fR with
the given ttl, which should enable udpcast to work between LANs. Not
tested though.
.IP "\-\-mcast\-all\-addr \fIaddress\fR" 4
.IX Item "--mcast-all-addr address"
Uses a non-standard multicast address for the control connection
(which is used by the sender and receivers to \*(L"find\*(R" each other). This
is \fBnot\fR the address that is used to transfer the data. By
default \f(CW\*(C`mcast\-all\-addr\*(C'\fR is the Ethernet broadcast address if
\&\f(CW\*(C`ttl\*(C'\fR is 1, and \f(CW224.0.0.1\fR otherwise. This setting
should not be used except in very special situations, such as when
\&\f(CW224.0.0.1\fR cannot be used for policy reasons.
.IP "\-\-exitWait \fImilliseconds\fR" 4
.IX Item "--exitWait milliseconds"
When transmission is over, receiver will wait for this time after
receiving the final \s-1REQACK\s0. This is done in order to guard against
loss of the final \s-1ACK\s0. Is 500 milliseconds by default.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
udp-sender
.SH "AUTHOR"
.IX Header "AUTHOR"
Alain Knaff


Multicast/udpcast-20040531/udp-receiver.c
#include <sys/types.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <getopt.h>
#include <signal.h>

#include "log.h"
#include "udpcast.h"
#include "udp-receiver.h"
#include "socklib.h"
#include "udpc_version.h"

#ifdef BB_FEATURE_UDPCAST_FEC
#include "fec.h"
#endif

#define MAXBLOCK 1024

static struct option options[] = {
    { "file", 1, NULL, 'f' },
    { "pipe", 1, NULL, 'p' },
    { "port", 1, NULL, 'P' },
    { "portbase", 1, NULL, 'P' },
    { "interface", 1, NULL, 'i' },
    { "ttl", 1, NULL, 0x0401 },
    { "mcast_all_addr", 1, NULL, 'M' },
    { "mcast-all-addr", 1, NULL, 'M' },

#ifdef LOSSTEST
    /* simulating packet loss */
    { "write-loss", 1, NULL, 0x601 },
    { "read-loss", 1, NULL, 0x602 },
    { "seed", 1, NULL, 0x603 },
    { "print-seed", 0, NULL, 0x604 },
    { "read-swap", 1, NULL, 0x605 },
#endif

    { "passive", 0, NULL, 0x0801 },

    { "nosync", 0, NULL, 0x0901 },

    { "rcvbuf", 1, NULL, 0x0a01 },

    { "nokbd", 0, NULL, 0xb04 },

    { "exitWait", 1, NULL, 0xd01 },

#ifdef BB_FEATURE_UDPCAST_FEC
    { "license", 0, NULL, 'L' },
#endif

    { "log", 1, NULL, 'l' },
};

static void intHandler(int nr) {
    udpc_fatal(1, "Cancelled by user\n");
}

#ifdef NO_BB
static void usage(char *progname) {
    fprintf(stderr, "%s [--file file] [--pipe pipe] [--portbase portbase] [--interface net-interface] [--log file] [--ttl time-to-live] [--mcast-all-addr mcast-all-address] [--rcvbuf buf] [--nokbd] [--exitWait milliseconds] [--license]\n", 
	    progname);
    exit(1);
}
#endif

#ifndef NO_BB
int udpreceiver_main(int argc, char **argv)
#else
int main(int argc, char **argv)
#endif
{
    char *ptr;
    struct net_config net_config;
    struct disk_config disk_config;
    int c;
    int doWarn=0;

#ifdef LOSSTEST
    int seedSet = 0;
    int printSeed = 0;
#endif

    disk_config.fileName=NULL;
    disk_config.pipeName=NULL;
    disk_config.flags = 0;

    net_config.portBase = 9000;
    net_config.rateLimit = NULL;
    net_config.ttl = 1;
    net_config.flags = 0;
    net_config.mcastAll = NULL;
    net_config.exitWait = 500;

#ifdef __CYGWIN__
    /* cygwin is basically unusable with its default buffer size of 8k...*/
    net_config.requestedBufSize = 1024*1024;
#else
    net_config.requestedBufSize = 0;
#endif

    ptr = strrchr(argv[0], '/');
    if(!ptr)
	ptr = argv[0];
    else
	ptr++;
    
    net_config.net_if = NULL;
    if (strcmp(ptr, "init") == 0) {
	doWarn = 1;
	disk_config.pipeName = strdup("/bin/gzip -dc");
	disk_config.fileName = "/dev/hda";
    }
    while( (c=getopt_long(argc, argv, "f:p:P:i:l:M:L",
			  options, NULL)) != EOF ) {
	switch(c) {
	    case 'f':
		disk_config.fileName=optarg;
		break;
	    case 'i':
		net_config.net_if=optarg;
		break;
	    case 'p':
		disk_config.pipeName=optarg;
		break;
	    case 'P':
		net_config.portBase = atoi(optarg);
		break;
	    case 'l':
		udpc_log = fopen(optarg, "a");
		break;
	    case 0x0401:
		net_config.ttl = atoi(optarg);
		break;
	    case 'M':
		net_config.mcastAll = strdup(optarg);
		break;

#ifdef BB_FEATURE_UDPCAST_FEC
	    case 'L':
		    fec_license();
		    break;
#endif

#ifdef LOSSTEST
	    case 0x601:
		setWriteLoss(optarg);
		break;
	    case 0x602:
		setReadLoss(optarg);
		break;
	    case 0x603:
		seedSet=1;
		srandom(strtoul(optarg,0,0));
		break;
	    case 0x604:
		printSeed=1;
		break;
	    case 0x605:
		setReadSwap(optarg);
		break;
#endif
	    case 0x801:
		net_config.flags|=FLAG_PASSIVE;
		break;
	    case 0x901:
		disk_config.flags|=FLAG_NOSYNC;
		break;
	    case 0xa01:
		net_config.requestedBufSize=parseSize(optarg);
		break;
	    case 0xb04:
		net_config.flags |= FLAG_NOKBD;
		break;

	    case 0xd01:
		net_config.exitWait = atoi(optarg);
		break;

	    case '?':
#ifndef NO_BB
	        bb_show_usage();
#else
		usage(argv[0]);
#endif
	}
    }

    fprintf(stderr, "Udp-receiver %s\n", version);

#ifdef LOSSTEST
    if(!seedSet)
	srandomTime(printSeed);
#endif

#ifndef __CYGWIN__
    signal(SIGINT, intHandler);
#endif
    return startReceiver(doWarn, &disk_config, &net_config);
}
